---
title: "ag project"
author: "Emily Hong"
date: "2024-04-17"
output: html_document
---

DESeq2 - Differential Analysis

```{r}
# install DESeq2 if it is not installed
if (!require("DESeq2")) install.packages("DESeq2")
```

```{r}
# load in counts.matrix and save as readcount
# row.names=1 to set the rownames for the dataframe as the first column
readcount = read.delim("counts.matrix", row.names = 1)
head(readcount)
```

```{r}
## create expgroup dataframe
# first 6 samples are WR, next 6 are ABA, and last 6 are Control
# set the Control as the reference level
expgroup = data.frame(condition = factor(c(rep("WR", 6), rep("ABA", 6), rep("Control", 6)), levels = c("Control", "WR", "ABA")))

# set the rownames of expgroup to the column names as readcount
row.names(expgroup) = colnames(readcount)
head(expgroup)
```

```{r}
# load DESeq2
library(DESeq2)
```

```{r}
## create a Counts Dataset Matrix using DESeqDataSetFromMatrix and name it cds
# countData: readcount dataframe, which includes count data
# colData: expgroup (rowname of expgroup corresponds to column name of readcount)
# design: ~ condition is a design formula that specifies that counts in the readcount dataframe for each gene depend on their condition in expgroup
cds = DESeqDataSetFromMatrix(countData = readcount, colData = expgroup, design = ~ condition)
head(cds)
```

```{r}
# filter out genes with low counts (less than or equal to 10)
cds = cds[rowSums(counts(cds)) > 10,]

# set control condition as reference
cds$condition <- relevel(cds$condition, ref = "Control")
```

```{r}
## use estimateSizeFactors and estimateDispersions on cds
# estimateSizeFactors: considers the difference in total reads from different samples and estimates the size factors for each gene
# estimateDispersions: estimates dispersion of counts for each gene
cds = estimateSizeFactors(cds)
cds = estimateDispersions(cds)
```

```{r}
# plot the per-gene dispersion in cds using plotDispEsts
plotDispEsts(cds)
```

```{r}
# run the DESeq function using cds
# obtain DESeq results using results function and call it res
cds = DESeq(cds)
```

```{r}
# save the results from DeSeq as res
res = results(cds)
head(res)
```

```{r}
# keep only the rows of res where padj is not NA
res <- res[!is.na(res$padj),]
head(res)
```

```{r}
## sum all of the genes that have an adjusted p-value of less than 0.05 and log2FoldC hange greater than 1 or less than -1, using | for or
# na.rm=T removes missing values (NA) in res
sum(res$padj < 0.05 & (res$log2FoldChange>1 | res$log2FoldChange < -1), na.rm=T)
```

```{r}
## save rownames of res where the adjusted p-value of less than 0.05 and log2FoldChan ge greater than 1 or less than -1 into list called diffexpgenes
diffexpgenes = row.names(res[res$padj < 0.05 & (res$log2FoldChange > 1 | res$log2FoldChange < -1), ])
```

```{r}
 # get normalized values of cds using normalized = T
normvalues = counts(cds, normalized = T)
head(normvalues)
```

```{r}
# subset normvalues so that only rownames of normvalues in diffexpgenes are kept
diffexpvalues = subset(normvalues, rownames(normvalues) %in% diffexpgenes)
```

```{r}
## create plot of the normalized expression values of the top 20 genes

# order DESeq results res by padj values
# then take only the rownames of the top 20 
res <- res[order(res$padj), ]
top_20_genes <- rownames(head(res, 20))

# get normalized expression values for the top 20 genes
norm_values20 <- normvalues[top_20_genes, ]

# set the colors of the conditions (Control, WR, and ABA)
condition_colors <- c("Control" = "lightgoldenrod", "WR" = "darkgreen", "ABA" = "tomato")

# set margins of the plot
par(mar = c(9, 6, 6, 4) + 0.1)

## initializes the plot with an empty plot containing a single point
# ylim = range(norm_values20): sets the y-axis limits based on the range of values in the norm_values20 vector
# xlim = c(1, length(top_20_genes)): sets the x-axis limits from 1 to the length of the top_20_genes vector
# xlab="", ylab = "Normalized Values": specifies the labels for the x-axis (xlab) and y-axis (ylab)
# main = "Normalized Expression Values of Top 20 Genes": specifies the title of the plot
# xaxt = "n": suppresses the x-axis ticks and labels; the xaxt argument is set to "n", indicating that no x-axis annotations should be drawn initially.

plot(1, type = "n", ylim = range(norm_values20), 
     xlim = c(1, length(top_20_genes)),
     xlab="",ylab = "Normalized Values",
     main = "Normalized Expression Values of Top 20 Genes",
     xaxt = "n")

## add dots for each gene, color coded by expgroup
# for (i in 1:length(top_20_genes)): iterates over the indices of the genes in the top_20_genes vector
# for (j in 1:ncol(norm_values20)): iterates over the columns of the norm_values20 matrix, which likely represent different conditions or experimental groups
# condition <- expgroup$condition[j]: extracts the condition (or experimental group) associated with the current gene (specified by the index j) from the expgroup dataframe
# points(i, norm_values20[i, j], pch = 16, col = condition_colors[condition]): plots a point on the plot
# pch = 16 specifies the point shape as a solid dot
#col = condition_colors[condition] specifies the color of the point based on the condition of the gene, using the condition_colors vector

for (i in 1:length(top_20_genes)) {
  for (j in 1:ncol(norm_values20)) {
    condition <- expgroup$condition[j]
    points(i, norm_values20[i, j], pch = 16, col = condition_colors[condition])
  }
}

# add legend for expgroup
legend("topright", legend = levels(expgroup$condition), fill = condition_colors, cex = 0.8)

# add x-axis labels
axis(side = 1, at = 1:length(top_20_genes), labels = top_20_genes, las = 2, cex.axis = 0.8)
```

```{r}
## create volcano plot
with(res, plot(log2FoldChange, -log10(pvalue), pch=20, main="Volcano plot", xlim=c(-8,8)))

# add colored points: blue if padj<0.05, red if log2FC>1 and padj<0.05)
with(subset(res, padj<.05 ), points(log2FoldChange, -log10(pvalue), pch=20, col="blue"))
with(subset(res, padj<.05 & abs(log2FoldChange)>1), points(log2FoldChange, -log10(pvalue), pch=20, col="red"))
```

Clustering

```{r}
## making cluster dendrogram
# diffexpvalues.cor = cor(t(diffexpvalues)): calculates the correlation matrix of the transposed gene expression data (diffexpvalues)
# diffexpvalues.dist= as.dist(1-diffexpvalues.cor): converts the correlation matrix into a distance matrix using the formula 1 - correlation
# diffexp.clust = hclust(diffexpvalues.dist, method="average"): performs hierarchical clustering on the distance matrix using the average linkage method
# diffexp.clust.groups=cutree(diffexp.clust, k=2): assigns each gene to a cluster based on the hierarchical clustering result (diffexp.clust), producing k=2 clusters

diffexpvalues.cor = cor(t(diffexpvalues))
diffexpvalues.dist= as.dist(1-diffexpvalues.cor) 
diffexp.clust = hclust(diffexpvalues.dist, method="average")
```

```{r}
# checking how many clusters we should use
# load cluster 
library(cluster)

# avg.sil.values=numeric(): initializes an empty numeric vector avg.sil.values to store the average silhouette values
# avg.sil.values[1]=0: sets the first element of the avg.sil.values vector to 0
# the loop iterates from i = 2 to i = 20
# temp.cluster.sil = silhouette(temp.clusters, dist=diffexpvalues.dist): calculates the silhouette width for each data point in the clustering result temp.clusters
# avg.sil.values[i]=mean(temp.cluster.sil[,"sil_width"]): calculates the average silhouette width for the current clustering result temp.clusters and stores it in the avg.sil.values vector

# plot(avg.sil.values): plots the average silhouette values against the number of clusters (k)

avg.sil.values=numeric()
avg.sil.values[1]=0
for (i in 2:20) {
  temp.clusters = cutree(diffexp.clust, k=i) 
  temp.cluster.sil = silhouette(temp.clusters, dist=diffexpvalues.dist)
  avg.sil.values[i]=mean(temp.cluster.sil[,"sil_width"])
}

plot(avg.sil.values)
```

```{r}
diffexp.clust.groups=cutree(diffexp.clust, k=2)

# table(diffexp.clust.groups): provides a count of the number of genes in each cluster
table(diffexp.clust.groups)
# creates a data frame (diffexp_clusters) containing the cluster assignments for each gene
diffexp_clusters= data.frame(diffexp.clust.groups)

# plotting cluster dendrogram
plot(diffexp.clust)
```

Heatmap

```{r}
# install pheatmap
if (!require("pheatmap")) install.packages("pheatmap")
```

```{r}
# load pheatmap
library(pheatmap)
```

```{r}
# create heatmap
pheatmap(diffexpvalues,
         annotation_col = expgroup,
         annotation_row= diffexp_clusters,
          cluster_rows = TRUE,
          cluster_cols = TRUE,
         show_rownames = FALSE,
         show_colnames= TRUE,
         scale="row",
         main="Heatmap for Differentially Expressed Genes in CON,ABA, and WR Subjects")
```

GO Term Analysis

```{r}
# load packages necessary for GO Term analysis
library(GOstats)
library(GO.db)
library(Category)
library(org.Mm.eg.db)
```

```{r}
# maps Ensembl gene IDs (specified by rownames(diffexpvalues)) to Entrez gene IDs using the org.Mm.eg.db annotation package for the Mus musculus (mouse) genome
# maps Ensembl gene IDs (specified by rownames(readcount)) to Entrez gene IDs using the same annotation package (org.Mm.eg.db)
entrez_ids = mapIds(org.Mm.eg.db, keys = rownames(diffexpvalues), keytype = "ENSEMBL", column = "ENTREZID")
universe_entrez_ids = mapIds(org.Mm.eg.db, keys = rownames(readcount), keytype = "ENSEMBL", column = "ENTREZID")
```

```{r}
# sets up parameters for performing Gene Ontology (GO) enrichment analysis using the hypergeometric test
params = new("GOHyperGParams",
       geneIds = entrez_ids,
       universeGeneIds = universe_entrez_ids,
       annotation = "org.Mm.eg.db",
       ontology = "BP",
       pvalueCutoff = 0.001,
       testDirection = "over")
```

```{r}
# performs the hypergeometric test for gene set enrichment analysis using the parameters specified in params and then summarizes the results
overRepresented = hyperGTest(params)
head(summary(overRepresented))
```

```{r}
## create a significant_terms data frame from overRepresented
significant_terms = as.data.frame(summary(overRepresented))
# order the significant_terms based on Pvalue
significant_terms = significant_terms[order(significant_terms$Pvalue), ]

# select the top 20 significant terms
top_20_terms = significant_terms[1:20, ]

# select relevant columns (GOBPID, Term, Pvalue, Count)
top_20_terms = top_20_terms[, c("GOBPID", "Term", "Pvalue", "Count")]

# display the top_20_terms data frame
top_20_terms
```

```{r}
# load ggplot 
library(ggplot2)
```

```{r}
## create a GO Terms Pathway Bubble Plot
ggplot(top_20_terms, aes(x = ifelse(Count, Count, "Others"), y = Term, size = Count, color = -log10(Pvalue))) +
  geom_point(alpha = 0.7) +
  scale_size_continuous(name = "Gene count", range = c(3, 7)) +
  scale_color_gradient(low = "blue", high = "red", name = "-log10(P-value)") +
  labs(title = "Top 20 GO-terms Pathways Bubble Plot",
       x = "Number of Genes",
       y = "GO Term Pathway",
       size = "Gene count",
       color = "-log10(P-value)") +
  theme_minimal() +
  theme(axis.text.y = element_text(hjust = 1))
```

PCA

```{r}
## calculate the principal components of the data stored in diffexpvalues
# transposes diffexpvalues using t()
# scale=TRUE argument: variables will be scaled to have unit variance before the analysis
# center=TRUE: variables will be centered (their mean will be subtracted from each variable)
data.prcomp = prcomp(t(diffexpvalues), scale=TRUE, center=TRUE)
head(data.prcomp$sdev) # shows the standard deviations of the principal components

# displays the rotation matrix, which shows the loadings of the original variables on the principal components
data.prcomp$rotation[1:6,1:6]
data.prcomp$x[1:6.1:6] # scores of the observations on the first six principal components

plot(data.prcomp) # plots the results of the principal component analysis
summary(data.prcomp) # summary of the principal component analysis results
```

```{r}
coords2draw = as.data.frame(data.prcomp$x) # converts the matrix of principal component scores into a data frame named coords2draw
coords2draw = cbind(coords2draw, expgroup) # binds the expgroup variable to the coords2draw data frame

## create PCA plot
# maps the color aesthetic to the condition column of the expgroup variable
ggplot(coords2draw, aes(x = PC1, y = PC2, color = expgroup$condition)) +
  geom_point() +
  theme_classic() +
  scale_color_manual(values = c("Control" = "blue", "WR" = "red", "ABA" = "darkgray")) +
  labs(color = "Condition", title = "PC1 vs PC2 Plot") +
  theme(plot.title = element_text(hjust = 0.5))
```
